internal/mq/msgstream/factory.go

```diff
@@ -0,0 +1,130 @@
+package msgstream
+
+import (
+	"context"
+	"fmt"
+
+	nmqimplserver "github.com/milvus-io/milvus/internal/mq/mqimpl/natsmq/server"
+	rmqimplserver "github.com/milvus-io/milvus/internal/mq/mqimpl/rocksmq/server"
+	"github.com/milvus-io/milvus/internal/mq/msgstream/mqwrapper/nmq"
+	"github.com/milvus-io/milvus/internal/mq/msgstream/mqwrapper/rmq"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/mq/msgstream"
+	"github.com/milvus-io/milvus/pkg/mq/msgstream/mqwrapper"
+	"go.uber.org/zap"
+)
+
+const (
+	// MsgStreamTypeRmq represents a message stream based on rocksdb.
+	MsgStreamTypeRmq StreamType = 1
+
+	// MsgStreamTypeNmq represents a message stream based on NATS-server.
+	MsgStreamTypeNmq StreamType = 2
+)
+
+var (
+	_ fmt.Stringer      = StreamType(1)
+	_ msgstream.Factory = &factory{}
+)
+
+// StreamType represents the type of message stream.
+type StreamType int
+
+func (t StreamType) String() string {
+	switch t {
+	case MsgStreamTypeRmq:
+		return "rmq"
+	case MsgStreamTypeNmq:
+		return "nmq"
+	default:
+		return ""
+	}
+}
+
+// Factory is a factory for creating message streams.
+//
+// It contains a function field named newer, which is a function that creates
+// an mqwrapper.Client when called.
+type factory struct {
+	streamType        StreamType
+	newer             func() (mqwrapper.Client, error) // client constructor
+	dispatcherFactory msgstream.ProtoUDFactory
+	receiveBufSize    int64
+	mqBufSize         int64
+}
+
+// NewFactory creates a new message stream factory of type t using the given path.
+func NewFactory(t StreamType, path string) msgstream.Factory {
+	var newer func() (mqwrapper.Client, error)
+	var err error
+	switch t {
+	case MsgStreamTypeRmq:
+		newer = rmq.NewClientWithDefaultOptions
+		err = rmqimplserver.InitRocksMQ(path)
+	case MsgStreamTypeNmq:
+		newer = nmq.NewClientWithDefaultOptions
+		err = nmqimplserver.InitNatsMQ(path)
+	default:
+		panic("unreachable msgstream type")
+	}
+	if err != nil {
+		// Should be stop if fail to create local msgstream service.
+		log.Fatal("init msgstream failed", zap.String("msgStreamType", t.String()), zap.Error(err))
+	}
+	log.Info("init local msgstream success", zap.String("msgStreamType", t.String()), zap.String("path", path))
+
+	return &factory{
+		streamType:        t,
+		newer:             newer,
+		dispatcherFactory: msgstream.ProtoUDFactory{},
+		receiveBufSize:    1024,
+		mqBufSize:         1024,
+	}
+}
+
+// StreamType returns the current stream type used by the factory.
+func (f *factory) StreamType() StreamType {
+	return f.streamType
+}
+
+// NewMsgStream is used to generate a new Msgstream object
+func (f *factory) NewMsgStream(ctx context.Context) (msgstream.MsgStream, error) {
+	cli, err := f.newer()
+	if err != nil {
+		return nil, err
```

Comment:
Thanks for properly handling the error.
Could you wrap more info into the error, i.e.. `return nil, util.WrapError("failed to create new msgstream", err)`.
In this file, there are multiple such errors from calling `f.newer()`. The caller doesn't necessarily log more info. Thus, I'd prefer adding more info into the log for each of them. 
It could be quite useful when we debug it.

Reply:
Great!
new commit changed:
Add a error wrapper as defer function for msg stream creator function.
implementation is moved to pkg/mq/msgstream/`CommonFactory`